docs/adding_an_e2e_test.md

```diff
@@ -0,0 +1,106 @@
+
+From the structure of the annotations for the arguments other than the `self` argument it is clear that only tensor arguments are supported. This means that if an op requires an input other than a tensor, it should be created in the method body or as a class parameter in the `__init__` method.
+
+### `forward` Method
+
+```python
+    def forward(self, x, index):
+        return torch.ops.aten.index(x, (index,))
+```
+
+The forward method should be a simple test of your op. In other words, it will almost always take the form of simply returning the result of calling your op. The call to your op should **always** be made using `torch.ops.aten.{op_name}` to make it very clear which ATen op is being tested. Some ATen ops have different variants under the same base name, such as `aten.mean`, which has also a variant `aten.mean.dim`. At the Python level, such ops are accessed by just their base name, and the right variant is chosen based on the inputs given. For example, to test `aten.mean.dim` the test should use `torch.ops.aten.mean(..., dim=...)`.
+
+### `@register_test_case` Decorator
+
+```python
+@register_test_case(module_factory=lambda: IndexTensorModule3dInput())
+```
+
+The `@register_test_case` decorator is used to register the test case function. The `module_factory` argument should be a function that when called produces an instance of the test class. This function will be used to create the first argument passed to the test case function.
+
+### Test Case Function
+
+```python
+def IndexTensorModule3dInput_basic(module, tu: TestUtils):
+    module.forward(tu.rand(5, 4, 3), torch.randint(3, (2, 3)))
+```
+
+The convention adopted for the name of the test case function is to have the same name as the test class postfixed by `_basic`. The test function always takes an instance of the test class as the first argument and a [`TestUtils`](https://github.com/llvm/torch-mlir/blob/ba17a4d6c09b4bbb4ef21b1d8d4a93cb056be109/python/torch_mlir_e2e_test/torchscript/framework.py#L167) object as the second argument. The `TestUtils` has some methods that allow the creation of random tensors. However, there is currently no support for generating random integer tensors, so if that is needed use `torch.randint`.
+
+
+## Things to Consider When Creating New Tests
+
+- Do you need negative numbers? If so, [`tu.rand`](https://github.com/llvm/torch-mlir/blob/2374098d71083bce0fd2b665e28ebb621de5779c/python/torch_mlir_e2e_test/torchscript/framework.py#L182) and [`torch.randint`](https://pytorch.org/docs/stable/generated/torch.randint.html) both allow you to specify a lower bound for random number generation
+- Do you need ints? Use `torch.randint` rather than `tu.rand(...).to(dtype=int)`
+- Make sure the annotation of the forward method matches the input types and ranks
```

Comment:
Can you file a bug for this? This should somehow fail if it is wrong.

Reply:
Here is the issue: https://github.com/llvm/torch-mlir/issues/1294